### DISCUSSION TITLE
How to get the FullToken or GlobalTokenWithComponent?

### DISCUSSION THREAD

**Discussion by thomastvedt** [ant-design/ant-design#39609] [Q&A]
Hello, I'm upgrading to v5. I have a custom menu component where I want to follow the same-ish design as ant design Menu component. 

If I run this:

```js
const { token } = useToken(); 
```

```js
console.log(token.Menu?.colorItemText); // => undefined
```

The token above is of type `type GlobalToken = AliasToken & ComponentTokenMap;`.

How can I consume / read the "complete" token with component styles?

```js
export type FullToken<ComponentName extends OverrideComponent> = TokenWithCommonCls<
  GlobalTokenWithComponent<ComponentName>
>;
```

**Comment by HyperLife1119** [comment]
This is a method I discovered, although it's not very elegant :(

```ts
const { prepareComponentToken } = await import('antd/es/button/style/token')

console.log(prepareComponentToken(theme.getDesignToken()))
```

**Comment by skyzhao1223** [comment]
`theme.useToken()` only ever returns the **global** token (Seed → algorithm → Map → Alias). It never contains component-level tokens like `Menu.colorItemText`. The type `GlobalToken = AliasToken & ComponentTokenMap` misleads you here: **every key of `ComponentTokenMap` is optional** (`Menu?: MenuComponentToken` in `components/theme/interface/components.ts`), so TypeScript happily accepts `token.Menu?.colorItemText` while at runtime `token.Menu` is simply not there.

At runtime, `token.Menu` only exists if **you** configured `theme.components.Menu` on a `ConfigProvider` — and even then it holds only your **raw override object**, not antd's computed Menu tokens:

```tsx
<ConfigProvider theme={{ components: { Menu: { itemColor: 'red' } } }}>
```
```ts
const { token } = theme.useToken();
token.Menu;                  // => { itemColor: 'red' }  — your raw config, nothing else
token.Menu?.colorItemText;   // => undefined
```

Menu's actual default tokens (`colorItemText`/`itemColor`, `itemSelectedColor`, `darkItemBg`, …) are computed lazily **inside the style-generation pipeline** of the Menu component and are never put into any React context. There is **no public API that returns the merged ("Full") component token** — in any v5 or v6 version.

### Where the merged token actually lives (source walk)

- `components/theme/index.tsx` — public `theme.useToken()` returns `{ theme, token, hashId }`; `token` is the cached global derivative token (`realToken`, raw values even in `cssVar` mode — `components/theme/useToken.ts`, `return [mergedTheme, realToken, ...]`).
- `getComputedToken` in `components/theme/useToken.ts` merges seed → algorithm → alias (`util/alias.ts`), then copies **only user-provided** `components.*` entries onto the token object verbatim. No component defaults are computed here.
- The merge you're looking for happens in `genComponentStyleHook` (in your early-v5 version: `components/theme/util/genComponentStyleHook.ts`; since antd 5.20.0 it lives in `@ant-design/cssinjs-utils`, antd's side is `components/theme/util/genStyleUtils.ts`). Inside the style function it does, in effect:

```ts
const defaultComponentToken = prepareComponentToken({ ...globalToken, ...(globalToken.Menu ?? {}) });
const componentToken = { ...defaultComponentToken, ...globalToken.Menu };   // your ConfigProvider overrides win
const fullToken = mergeToken(globalToken, { componentCls, prefixCls, iconCls, antCls, ... }, componentToken);
// fullToken is passed ONLY to Menu's style functions — this is the `FullToken<'Menu'>` you found
```

That `fullToken` is a transient object created per style registration. It is not stored in context, so nothing can hand it to you. The `FullToken` / `TokenWithCommonCls` types you quoted describe that internal object (including styling plumbing like `componentCls`, `iconCls`, `antCls`) — they were never meant to be obtainable at runtime by consumers.

### Recommended solution: replicate antd's own computation (verified)

Each component's style module exports the function that produces its default component token: `prepareComponentToken` (Menu: `components/menu/style/index.ts`; Button: `components/button/style/token.ts`). Combine it with `theme.useToken()` exactly the way antd does internally:

```ts
// useMenuToken.ts
import { theme } from 'antd';
import type { GlobalToken } from 'antd';
import { prepareComponentToken } from 'antd/es/menu/style';
import type { ComponentToken as MenuComponentToken } from 'antd/es/menu/style';

type MenuToken = GlobalToken & MenuComponentToken;

// `prepareComponentToken` is typed as the internal `GetDefaultToken<'Menu'>`
// (a union of object | function | null), so cast it to its actual function form:
const getMenuTokenDefaults = prepareComponentToken as unknown as (
  token: GlobalToken & Partial<MenuComponentToken>,
) => MenuComponentToken;

export function useMenuToken(): MenuToken {
  const { token } = theme.useToken();
  // `token.Menu` carries the raw `theme.components.Menu` overrides from the
  // nearest ConfigProvider (undefined when none are set)
  const override: Partial<MenuComponentToken> = token.Menu ?? {};
  const defaults = getMenuTokenDefaults({ ...token, ...override });
  return { ...token, ...defaults, ...override };
}
```

```tsx
const menuToken = useMenuToken();
menuToken.colorItemText;      // 'rgba(0,0,0,0.88)'  (default light theme)
menuToken.itemColor;          // 'rgba(0,0,0,0.88)'  (current name of the same token)
menuToken.itemSelectedColor;  // '#1677ff'
menuToken.darkItemColor;      // 'rgba(255,255,255,0.65)'
menuToken.colorPrimary;       // all global tokens are included too
```

I verified this hook by actually running it (Node SSR, antd **5.29.3** + React 18, and antd **6.6.5** + React 19). Its output matches, key for key, the CSS variables antd itself registers for Menu in `cssVar` mode (`--ant-menu-item-color`, `--ant-menu-item-selected-color`, `--ant-menu-item-bg`, …) under: default theme, `components.Menu` overrides, global `token` overrides, `darkAlgorithm`, and per-component `components.Menu.algorithm`. Sample results:

| Scenario | `itemColor` | `itemSelectedColor` | `itemBg` |
| --- | --- | --- | --- |
| default | `rgba(0,0,0,0.88)` | `#1677ff` | `#ffffff` |
| `components.Menu.itemColor: 'rgb(255,0,0)'` | `rgb(255,0,0)` | `#1677ff` | `#ffffff` |
| `algorithm: darkAlgorithm` | `rgba(255,255,255,0.85)` | `#1668dc` | `#141414` |

**Caveats — please read:**

1. `prepareComponentToken` is an **internal export**, not part of antd's documented API. antd's `package.json` has no `exports` field, so `antd/es/menu/style` deep imports physically resolve, but they carry no semver guarantee. History shows these paths move: the export didn't exist before 5.12.0 (Menu) / 5.13.0 (Button), and the surrounding pipeline moved to `@ant-design/cssinjs-utils` in 5.20.0. It exists and works in every current 5.x (verified 5.29.3) and in 6.x (verified 6.6.5), but pin your antd version and re-check on upgrades.
2. If you override a **deprecated** token name (e.g. `components.Menu.colorItemText`), antd internally maps it to the new name (`itemColor`) via its `deprecatedTokens` table; the simple spread above does not. Override using current token names.
3. The returned object contains global + component token **values** only. The internal styling helpers (`componentCls`, `iconCls`, `antCls`, `calc`) from `FullToken` are pipeline plumbing and are not included — you don't need them to style a custom component.

### If you want documented public API only

There is no public API that returns merged component tokens, so the 100%-public option is to derive the few values you need from the global token yourself. Menu's defaults are exactly these mappings (source: `prepareComponentToken` in `components/menu/style/index.ts`):

```ts
const { token } = theme.useToken();
const myMenuToken = {
  itemColor: token.colorText,
  itemHoverColor: token.colorText,
  itemSelectedColor: token.colorPrimary,
  itemBg: token.colorBgContainer,
  itemHoverBg: token.colorBgTextHover,
  itemSelectedBg: token.controlItemBgActive,
  subMenuItemBg: token.colorFillAlter,
  itemBorderRadius: token.borderRadiusLG,
  itemHeight: token.controlHeightLG,
  popupBg: token.colorBgElevated,
  groupTitleColor: token.colorTextDescription,
  itemDisabledColor: token.colorTextDisabled,
  // dark theme (Menu theme="dark"):
  darkItemBg: '#001529',
  darkSubMenuItemBg: '#000c17',
  darkItemColor: /* colorTextLightSolid @ 65% alpha */ 'rgba(255,255,255,0.65)',
  darkItemSelectedColor: token.colorTextLightSolid,
  darkItemSelectedBg: token.colorPrimary,
  // user overrides from ConfigProvider, if any:
  ...(token.Menu ?? {}),
};
```

The trade-off: you own this mapping and must maintain it across antd versions. Also useful:

- `theme.getDesignToken(config?)` (public since 5.6.0) — same global token, statically, outside React. It accepts `{ token, algorithm }`, but note it **ignores `config.components`** entirely (implementation: `components/theme/getDesignToken.ts` only reads `config.token`/`config.algorithm`).
- `ConfigProvider.useConfig()` does **not** expose theme data — it returns only `{ componentDisabled, componentSize }` (`components/config-provider/hooks/useConfig.ts`). The raw merged `ThemeConfig` (including your `components.Menu` overrides) is readable from `React.useContext(ConfigProvider.ConfigContext).theme`, but that gives you user config, not computed tokens; `token.Menu` from `useToken()` carries the same overrides.

### About the existing `prepareComponentToken` reply

The approach in the earlier reply is directionally correct — that is indeed the function antd uses — with these corrections:

1. **The path is Button-specific.** `antd/es/button/style/token` exists (since 5.13.0) and does export `prepareComponentToken`. For **Menu** there is no `style/token` module; it's `import { prepareComponentToken } from 'antd/es/menu/style'` (since 5.12.0 — before that it was an anonymous inline function, not exported at all).
2. **`getDesignToken()` gives you the default theme only.** `prepareComponentToken(theme.getDesignToken())` ignores whatever `ConfigProvider` theme is active where your component renders. Inside React, base it on `theme.useToken().token` (the hook above); statically, pass the same config you pass to `ConfigProvider` and merge component overrides yourself: `prepareComponentToken({ ...theme.getDesignToken(cfg), ...cfg?.components?.Menu })`.
3. It returns **only** the component-level keys — not the merged global+component token — so spread it over the global token as shown above.
4. Same caveat as above: it's an internal deep import, not covered by semver.

### Version notes

- `colorItemText` (the token in your question) was **renamed to `itemColor` in 5.13.0**; the old name is deprecated but still computed and still accepted in `components.Menu` overrides (via antd's `deprecatedTokens` mapping). Both exist in current 5.x and 6.x.
- The file you quoted (`FullToken` in `genComponentStyleHook.ts`) reflects early v5. Since 5.20.0 that logic lives in `@ant-design/cssinjs-utils`; antd re-exports the types (`FullToken`, `GetDefaultToken`, `ComponentTokenMap`, …) from `antd/es/theme/interface` (type-only deep import; from the `antd` root only `GlobalToken` and `MappingAlgorithm` are exported).

I ran everything above under Node SSR on antd 5.29.3 and then again on 6.6.5 (React 18 and 19 respectively) rather than only type-checking it.
